#include "UpgradeMgr.h"
#include "QQmlContext"
#include "QQuickView"
#include <QDebug>
#include <iostream>
#include <string>
// #include <unistd.h>

#define LOGGER GetLogMgr().getLogger()

UpgradeMgr::UpgradeMgr(QObject *parent) : QObject{parent} {
  // TODO urlä¸ºTboxæœåŠ¡ç«¯çš„ipåœ°å€
  cmd_server_url = "tcp://192.168.1.1:9529";
}

inline CLogMgr& UpgradeMgr::GetLogMgr() {
  static CLogMgr my_logger("MAPP", "MCON", E_FUNC_LINE, DLT);
  return my_logger;
}

UpgradeMgr& UpgradeMgr::getInstance() {
  static UpgradeMgr s_upgradeMgr;
  return s_upgradeMgr;
}

void UpgradeMgr::init() {
  try
  {
    m_ZmqClient = std::make_unique<ZmqClient>(
    cmd_server_url, [this](const std::string &msg) { handle_callback(msg); });
    // TODO éœ€è¦æœåŠ¡ç«¯æä¾›url
    m_zmqMonitorCb = std::bind(&UpgradeMgr::client_monitor,this,std::placeholders::_1,
    std::placeholders::_2);
    m_ZmqClient->set_monitor_callback(m_zmqMonitorCb); 
    bool connected = m_ZmqClient->connect(); 
    LOGGER->Cowa_Info(("connect to server, result = " + std::to_string(connected)).c_str());
  }
  catch(const std::exception& e)
  {
    LOGGER->Cowa_Error(e.what());
  }
  
}

void UpgradeMgr::handle_callback(const std::string &msg) {
  Command cmd;
  if (!cmd.ParseFromString(msg)) {
    LOGGER->Cowa_Warn("proto parse command failed!");
    return;
  }
  std::string out_msg = "";
  out_msg.append("receive msg from server, CmdId = ").append(std::to_string(cmd.id()))
         .append(", device = ").append(cmd.device())
         .append(", module = ").append(cmd.module())
         .append(", data = ").append(cmd.data());
  LOGGER->Cowa_Info(out_msg.c_str());
  switch (cmd.id()) {
  case CMD_ID_NOTIFY_DOWNLOAD_VERSIONS:
    // ä¸»å‡çº§ç¨‹åºå‘hmié€šçŸ¥æœ‰æ–°ç‰ˆæœ¬ï¼Œæ˜¯å¦éœ€è¦å‡çº?
    push_data_to_hmi(cmd.data());
    break;
  case CMD_ID_REQ_DOWNLOAD_VERSIONS:
    //è¯·æ±‚å‡çº§ä¸‹è½½ä¿¡æ¯
    // send_data_to_hmi(cmd.data());
    break;
  case CMD_ID_CONFIRM_INSTALL:
    // hmiç¡®è®¤æ˜¯å¦éœ€è¦å‡çº?
    // TODO é€šçŸ¥æœåŠ¡ç«¯å¼€å§‹å‡çº§ï¼Œhmiå…ˆä¸åšå¤„ç?
    LOGGER->Cowa_Info("notify server to install. ");
    break;
  default:
    LOGGER->Cowa_Warn("unknown command!");
    break;
  }
}

bool UpgradeMgr::request_identify() {
  // 1ã€æ„é€ è¯·æ±‚ç»“æ„Command
  // CmdId: CMD_ID_NOTIFY_IDENTITY, device: "cm4", module: "tboxhmi"
  Command cmd = create_command(CMD_ID_NOTIFY_IDENTITY, "cm4", "tboxhmi", "");
  // 2ã€åºåˆ—åŒ–ä¸ºstringç±»å‹
  std::string str_msg = cmd.SerializeAsString();
  // 3ã€é€šè¿‡ZmqClient.send()å‘é€msg
  bool result = m_ZmqClient->send(str_msg);
  LOGGER->Cowa_Info(("send identify to server, result = " + std::to_string(result)).c_str());
  return result;
}

Command UpgradeMgr::create_command(const CmdId& id, const std::string& device, const std::string& module, const std::string& data) {
  std::string out_msg = "";
  out_msg.append("create command, id = ").append(std::to_string(id))
         .append(", device = ").append(device)
         .append(", module = ").append(module)
         .append(", data = ").append(data);
  LOGGER->Cowa_Info(out_msg.c_str());
  Command cmd;
  cmd.set_id(id);
  cmd.set_device(device);
  cmd.set_module(module);
  cmd.set_data(data);
  return cmd;
}

void UpgradeMgr::push_data_to_hmi(const std::string &data) {
  // 1ã€è§£ææ¥æ”¶åˆ°çš„æ•°æ?
  InstallPackages install_pkgs;
  int pkgs_size = 0;
  std::shared_ptr<VersionInfo> info;
  install_pkgs.ParseFromString(data);
  // å¤§åŒ…å½“å‰ç‰ˆæœ¬
  QString curr_version = QString::fromStdString(install_pkgs.version_cur());
  // å¤§åŒ…æœ€æ–°ç‰ˆæœ?
  QString new_version = QString::fromStdString(install_pkgs.version_new());
  Q_EMIT signalNotifyMajorVersionInfo(curr_version, new_version);
  // å¾…å‡çº§çš„åŒ…æ•°é‡?
  pkgs_size = install_pkgs.pkgs_size();
  LOGGER->Cowa_Info(("push data to hmi, pkgs_size = " + std::to_string(pkgs_size)).c_str());
  // æ¸…ç©ºlist
  pkgs_list.clear();
  // 2ã€æ•°æ®å­˜å…¥listä¸?
  for (int i = 0; i < pkgs_size; i++) {
    info = std::make_shared<VersionInfo>();
    PackageUpdate updates = install_pkgs.pkgs(i);
    info->name = QString::fromStdString(updates.name());
    info->old_version = QString::fromStdString(updates.old_version());
    info->new_version = QString::fromStdString(updates.new_version());
    std::string out_msg = "";
    out_msg.append("updates.name: ").append(updates.name())
           .append(", updates.old_version: ").append(updates.old_version())
           .append(", updates.new_version: ").append(updates.new_version());
    LOGGER->Cowa_Info(out_msg.c_str());
    // é‡å¤æ•°æ®å¤„ç†ï¼Œç›¸åŒåº”ç”¨ï¼Œç›¸åŒç‰ˆæœ¬ä¸åœ¨HMIæ˜¾ç¤ºï¼Œä¸å‡çº§
    // if(!hasContains(info->name, info->old_version, info->new_version)) {
    //   pkgs_list.push_back(info);
    // }
    pkgs_list.push_back(info);
  }
  // æ›´æ–°UIç«?
  update_hmi_ui();
  // 3ã€æ˜¾ç¤ºå®‰è£…ç»“æœï¼šå‡çº§ä¸­ã€å‡çº§æˆåŠŸã€å‡çº§å¤±è´?
  InstallState state = install_pkgs.state();
  Q_EMIT signalNotifyUpgradeStatus(state);
}

// QString UpgradeMgr::StateToString(const InstallState& state) {
//   QString q_state_str;
//   qint32 q_state;
//   switch(state) {
//     case INSTALL_STATE_NORMAL:
//       // å®‰è£…å‰çš„çŠ¶æ€?
//       LOGGER->Cowa_Info("INSTALL_STATE_NORMAL");
//       break;
//     case INSTALL_STATE_PROCESSING:
//       // æ­£åœ¨å®‰è£…ä¸?
//       LOGGER->Cowa_Info("INSTALL_STATE_PROCESSING");
//       // q_state_str = "ç‰ˆæœ¬æ­£åœ¨å‡çº§ä¸?... ";
//       q_state = INSTALL_STATE_PROCESSING;
//       break;
//     case INSTALL_STATE_FINISH:
//       // å®‰è£…æˆåŠŸ
//       LOGGER->Cowa_Info("INSTALL_STATE_FINISH");
//       // q_state_str = "å‡çº§æˆåŠŸï¼?";
//       q_state = INSTALL_STATE_FINISH;
//       break;
//     case INSTALL_STATE_ERROR:
//       // å®‰è£…å¤±è´¥
//       LOGGER->Cowa_Info("INSTALL_STATE_ERROR");
//       // q_state_str = "å‡çº§å¤±è´¥ï¼?";
//       q_state = INSTALL_STATE_ERROR;
//       break;
//     case INSTALL_STATE_UNKNOW:
//       // æœªçŸ¥çŠ¶æ€?
//       // q_state_str = "æœªçŸ¥é”™è¯¯ï¼?";
//       q_state = INSTALL_STATE_UNKNOW;
//       LOGGER->Cowa_Info("INSTALL_STATE_UNKNOW");
//       break;
//     default:
//       break;
//   }
//   return q_state_str;
// }

void UpgradeMgr::update_hmi_ui() {
  std::list<std::shared_ptr<VersionInfo>>::iterator it;
  // å…ˆé€šçŸ¥hmiç«¯æ¸…ç©ºæ—§æ•°æ®
  Q_EMIT signalClearData();
  for (it = pkgs_list.begin(); it != pkgs_list.end(); it++)
  {
    std::shared_ptr<VersionInfo> info = *it;
    // å°†æ•°æ®ä¸»åŠ¨æ¨é€åˆ°qmlä¸?
    Q_EMIT signalNotifyPkgInfo(info->name, info->old_version, info->new_version);
  }
}

bool UpgradeMgr::req_server() {
  // 1ã€æ„é€ è¯·æ±‚ç»“æ„Command
  // CmdIdï¼šCMD_ID_REQ_DOWNLOAD_VERSIONSï¼Œè¯·æ±‚å‡çº§ä¸‹è½½ä¿¡æ?
  // device: "cm4"
  // module: "tboxhmi"
  // dataï¼?"
  Command cmd = create_command(CMD_ID_REQ_DOWNLOAD_VERSIONS, "cm4", "tboxhmi", "");
  // 2ã€åºåˆ—åŒ–ä¸ºstringç±»å‹
  std::string str_msg = cmd.SerializeAsString();
  // 3ã€é€šè¿‡ZmqClient.send()å‘é€msg
  bool result = m_ZmqClient->send(str_msg);
  LOGGER->Cowa_Info(("send request to server, result = " + std::to_string(result)).c_str());
  // TODO ä½¿ç”¨å‡æ•°æ®æµ‹è¯•å»é‡?
  // push_mock_data();
  return result;
}

void UpgradeMgr::push_mock_data() {
  std::cout << "push mock data." << std::endl;
  int size = 3;
  std::shared_ptr<VersionInfo> info = std::make_shared<VersionInfo>();
  info->name = QString::fromStdString("sau");
  info->old_version = QString::fromStdString("V1.0.1");
  info->new_version = QString::fromStdString("V1.0.1");
  if(addOrUpdateList(info->name, info->old_version, info->new_version)) {
      pkgs_list.push_back(info);
  }
  for (int i = 0; i < size; i++) {
    std::shared_ptr<VersionInfo> info = std::make_shared<VersionInfo>();
    info->name = QString::fromStdString("hmi_" + std::to_string(i));
    info->old_version = QString::fromStdString("V1.0." + std::to_string(i));
    info->new_version = QString::fromStdString("V2.0." + std::to_string(i));
    std::string out_msg = "";
    out_msg.append("name: ").append(info->name.toStdString())
           .append(", old_version: ").append(info->old_version.toStdString())
           .append(", new_version: ").append(info->new_version.toStdString());
    LOGGER->Cowa_Info(out_msg.c_str());
    if(addOrUpdateList(info->name, info->old_version, info->new_version)) {
      pkgs_list.push_back(info);
    }
  }
  update_hmi_ui();
  // é—´éš”å‘é€ä¿¡å?
  // sleep(3);
}

bool UpgradeMgr::addOrUpdateList(QString &name, QString &oldVersion, QString &newVersion) {
  std::list<std::shared_ptr<VersionInfo>>::iterator it;
  for (it = pkgs_list.begin(); it != pkgs_list.end(); it++)
  {
    std::shared_ptr<VersionInfo> info = *it;
    if(info->name == name) {
      info->old_version = oldVersion;
      info->new_version = newVersion;
      info->old_version.toStdString();
      std::string out_msg = "";
      // out_msg.append("*****old info******: ").append("old_version: ").append(info->old_version.toStdString())
      //      .append(", new_version: ").append(info->new_version.toStdString());
      // out_msg.append("*****new info******: ").append("now_old_version: ").append(oldVersion.toStdString())
      //      .append(", now_new_version: ").append(newVersion.toStdString());
      LOGGER->Cowa_Info("æ›´æ–°æ•°æ®");
      return false;
    }
  }
  return true;
}

bool UpgradeMgr::notify_server() {
  // 1ã€æ„é€ è¯·æ±‚ç»“æ„Command
  // CmdIdï¼šCMD_ID_CONFIRM_INSTALLï¼Œhmiç¡®è®¤æ˜¯å¦éœ€è¦å‡çº?
  // device: "cm4"
  // module: "tboxhmi"
  // dataï¼šInstallNotify
  InstallNotify install_notify;
  install_notify.set_update(1);
  Command cmd = create_command(CMD_ID_CONFIRM_INSTALL, "cm4", "tboxhmi", install_notify.SerializeAsString());
  // 2ã€åºåˆ—åŒ–ä¸ºstringç±»å‹
  std::string str_msg = cmd.SerializeAsString();
  // 3ã€é€šè¿‡ZmqClient.send()å‘é€msg
  bool result = m_ZmqClient->send(str_msg);
  LOGGER->Cowa_Info(("notify server to install, result = " + std::to_string(result)).c_str());
  // TODO æ¨¡æ‹ŸæœåŠ¡ç«¯çš„å›è°ƒç»“æœ
  // if(result) {
  //   // é€šçŸ¥hmiç«¯å‡çº§çŠ¶æ€?
  //   signalNotifyUpgradeStatus(2);
  // }
  return result;
}

void UpgradeMgr::client_monitor(int events, const char *addr) {
  bool connected;
  switch (events)
  {
  case ZMQ_EVENT_CONNECTED:
    // è¿æ¥æˆåŠŸï¼Œåˆ™å‘é€è®¤è¯ä¿¡æ?
    connected = request_identify();
    if(connected) {
      LOGGER->Cowa_Info("identify success. ");
    } else {
      LOGGER->Cowa_Warn("identify failed, please try again!");
    }
    break;
  case ZMQ_EVENT_DISCONNECTED: 
    // æœªè¿æ¥ï¼Œåˆ™é‡æ–°è¿æ?
    LOGGER->Cowa_Warn("failed to connect to server, please try again!");
    // m_ZmqClient->connect();
    break;
  case ZMQ_EVENT_CLOSED:
    LOGGER->Cowa_Warn("monitor is closed, please try again!");
    // å…³é—­è¿æ¥
    // m_ZmqClient->stop();
    break;
  default:
    LOGGER->Cowa_Warn("undefined event!");
    break;
  }
}

std::list<std::shared_ptr<VersionInfo>> UpgradeMgr::getPackageLists() {
  return pkgs_list;
}

void UpgradeMgr::destroy() {
    if(m_ZmqClient->connected()) {
      m_ZmqClient->stop();
    }
}